include/arch/x86/arch.h

```diff
@@ -530,12 +531,112 @@ extern FUNC_NORETURN void _SysFatalErrorHandler(unsigned int reason,
 
 
 #ifdef CONFIG_X86_STACK_PROTECTION
+extern struct task_state_segment _main_tss;
+
+#ifdef CONFIG_X86_USERSPACE
+void _x86_drop_to_user_mode(void);
+
+static inline u32_t _arch_syscall_invoke5(u32_t call_id, u32_t arg1, u32_t arg2,
+					  u32_t arg3, u32_t arg4, u32_t arg5)
+{
+	u32_t ret;
+
+	__asm__ volatile("int %[vector]"
+			 : "=a" (ret)
+			 : "a" (call_id), "d" (arg1), "c" (arg2),
+			   "b" (arg3), "D" (arg4), "S" (arg5),
+			   [vector] "i" (CONFIG_X86_SYSCALL_VECTOR));
+	return ret;
+}
+
+static inline u32_t _arch_syscall_invoke4(u32_t call_id, u32_t arg1, u32_t arg2,
+					  u32_t arg3, u32_t arg4)
```

Comment:
Another efficiency point: the C side of this marshalls the inbound function by putting an extra argument on the **front** of the list.  So the original (iamcu) function took "call_id" in EAX, but now needs to write it to EDX needlessly.  If you put call_id at the end (and fix up the entry code on the kernel side accordingly), the compiler can see that the arguments can stay in the registers they're already in.

Reply:
I'm pretty sure call_id stays in EAX...double checking thus

Reply:
yes it stays in EAX.
"a" (call_id) means the argument goes in EAX.

https://gcc.gnu.org/onlinedocs/gcc/Machine-Constraints.html#Machine-Constraints

Reply:
Sorry, I wrote that complaint wrong: the outer function would have had **arg1** in EAX, but we need to move it to EDX, right?

Reply:
no we don't. why do you think it needs to be moved to EDX? I'm very confused.
_x86_syscall_entry_stub expects call_id in EAX.
Nothing gets moved.

Reply:
you seem to be confusing call_id and arg1

Reply:
Maybe I'm misunderstanding the intent of the API?  I'm assuming you've got some call that looks like:

      int k_my_call(int my_arg);

And are wrapping it with something that looks like:

    int k_my_call(int my_arg)
    {
        _arch_syscall_invoke1(MY_SYSCALL_NUM, my_arg);
    }

In this case you'd arrive in k_my_call with my_arg in EAX, which the compiler will then need to move to EDX.  If MY_SYSCALL_NUM was the *last* argument, you wouldn't have this problem and the compiler would leave EAX unmolested.

Reply:
aha, *now* I see it. OK let me give that a try, thanks for bearing with me

Reply:
Would the cost of having an indirect call be too high?

```C
static const void (*syscall_table)()[] = {
    [SYS_FOO] = _sys_foo,
    [SYS_BAR] = _sys_bar,
    [SYS_BAZ] = _sys_baz,
};

static inline long _arch_syscall_invoke1(int num, u32_t arg)
{
    return syscall_table[num](arg);
}

static inline long _arch_syscall_invoke2(int num, u32_t arg1, u32_t arg2)
{
    return syscall_table[num](arg1, arg2);
}
```

(Error checking omitted.)

These being inline will most likely translate to a `lea` + `jmp *` if the compiler is smart enough to figure out this is actually a tail call. (This usually happens on x86, but not on xtensa.)

Reply:
Maybe I've made the comment in the wrong place -- yeah, this code wouldn't run in user land. It would be inside the interrupt trap, and `num` would have to be validated to avoid userland calling whatever it wants inside the kernel.

Reply:
i don't see a good way to validate the function pointer without adding it to the whole gperf contraption, which would waste memory since we wouldn't need a struct _k_object associated with it. i think it's simpler to just have a (probably generated table) mapping call IDs to function pointers.

Reply:
You don't need to use gperf here: that's precisely what's the `syscall_table` thing in the thing I wrote is: an array mapping the system call number to a function pointer. It's easy to validate the number: just check the bounds.

Reply:
we're in violent agreement then :)